internal/runner/service.go

```diff
@@ -563,58 +563,51 @@ func runnerWinsizeToPty(winsize *runnerv1.Winsize) *pty.Winsize {
 	}
 }
 
-func (r *runnerService) ResolveVars(ctx context.Context, req *runnerv1.ResolveVarsRequest) (*runnerv1.ResolveVarsResponse, error) {
-	// Add explicitly passed env as a source.
-	sources := []commandpkg.EnvResolverSource{
-		commandpkg.EnvResolverSourceFunc(req.Env),
-	}
-
-	// Add project env as a source.
-	proj, err := ConvertRunnerProject(req.Project)
-	if err != nil {
-		return nil, err
-	}
+func (r *runnerService) ResolveProgram(ctx context.Context, req *runnerv1.ResolveProgramRequest) (*runnerv1.ResolveProgramResponse, error) {
+	r.logger.Info("running ResolveProgram in runnerService")
+	resolver, err := r.getProgramResolverFromReq(req)
 	if err != nil {
 		return nil, err
 	}
-	if proj != nil {
-		projEnvs, err := proj.LoadEnv()
-		if err != nil {
-			r.logger.Info("failed to load envs for project", zap.Error(err))
-		} else {
-			sources = append(sources, commandpkg.EnvResolverSourceFunc(projEnvs))
-		}
-	}
-
-	// Add session env as a source.
-	session, found := r.getSessionFromRequest(req)
-	if found {
-		sources = append(sources, commandpkg.EnvResolverSourceFunc(session.Envs()))
-	}
-
-	resolver := commandpkg.NewEnvResolver(sources...)
 
-	var result []*commandpkg.EnvResolverResult
+	var varRes []*commandpkg.ProgramResolverResult
+	var scriptRes bytes.Buffer
 
 	if script := req.GetScript(); script != "" {
-		result, err = resolver.Resolve(strings.NewReader(script))
-	} else if commands := req.GetCommands(); commands != nil && len(commands.Items) > 0 {
-		result, err = resolver.Resolve(strings.NewReader(strings.Join(commands.Items, "\n")))
+		varRes, err = resolver.Resolve(strings.NewReader(script), &scriptRes)
+	} else if commands := req.GetCommands(); commands != nil && len(commands.Lines) > 0 {
+		varRes, err = resolver.Resolve(strings.NewReader(strings.Join(commands.Lines, "\n")), &scriptRes)
 	} else {
 		err = status.Error(codes.InvalidArgument, "either script or commands must be provided")
 	}
 	if err != nil {
 		return nil, err
 	}
 
-	response := &runnerv1.ResolveVarsResponse{}
+	response := &runnerv1.ResolveProgramResponse{
+		Commands: &runnerv1.ResolveProgramCommandList{
+			Lines: strings.Split(scriptRes.String(), "\n"),
```

Comment:
Using `\n` as a separator is not safe when we want to reconstruct the command list back. Within the script there might be escaped `\n` which here will not be properly handled. Likely the `Resolve` method need to work with a script and individual lines, which all should be valid shell commands, right?

Reply:
Yep, you're right if, eg a heredoc was gonna be inside of the script the splitting would garble it.

The reason why I'd like to return a list of commands here is to make it explicit in the API that normalization/transformation from script->commands has already happened. However, I'm clearly not handling this well.

Feel free to propose a solution @adambabik.

Reply:
Interestingly, it actually works! Please see: https://us-central1.stateful.com/cell/clt0fkcs6000288eiz1b9nbnu

I believe the reason is because the splitting happens on a string-level as opposed to the bytes where `\\n` vs `\n` are unescaped appropriately. Wdyt @adambabik?

Reply:
This is an interesting case but I don't quite understand how Go processes escaped `\n` in this case. 

I was thinking about a bit simpler example, however, I am not sure if this is how the client will use this method. The current definition allows such usage, though. 

Check out [this test](https://github.com/stateful/runme/pull/513/files#diff-5c90f0532937e162b42092e036a66d065a0118edae2ce20350e53dfea2ebf52bR85-R131).

Basically, with a program as follows:
```
export FILE_NAME=default.txt
cat >"$FILE_NAME" <<EOF
Some content with\nnew line
EOF
```

I expect the output to be:
```
#
# FILE_NAME set in smart env store
# "export FILE_NAME=default.txt"

cat >"$FILE_NAME" <<EOF
Some content with\nnew line
EOF
```

Instead, I got (+/- new line at the end):
```
#
# FILE_NAME set in smart env store
# "export FILE_NAME=default.txt", 

cat >"$FILE_NAME" <<EOF
Some content with
new line
EOF
```

Also, the question is why `ResolveProgramResponse` does not contain the source in the same format as in the input but always return a list of commands. Should that be fixed?

Reply:
Yes, you're right. You managed to "reproduce" the problem. The only thing I can think of is that on the extension side `\n` is being escaped as `\\n`. I can x-check.

@adambabik perhaps it'd be better to use the [SingleLine](https://pkg.go.dev/mvdan.cc/sh/v3@v3.8.0/syntax#SingleLine) option and break the commands at `;` instead?

Reply:
<img width="778" alt="image" src="https://github.com/stateful/runme/assets/250527/be5d9454-0998-437e-be5f-5833ca2f09cc">

@adambabik looks like the `\\n` is why it "works" ☝️

Reply:
Here's the same buf encoded as string printed with quotes (`%q`):

```
"cat >somefile <<EOF\nline1\nline2\\n\\n\nline3\nEOF\ncat somefile\n"
```

Reply:
In terms of the solution. Currently, the service accepts two kinds of program source: (1) a script as a single blob, (2) a list of commands. My first question would be whether we need both and in which cases the client uses one vs another.

The case of a script as a blob is easy, provided that the service returns it as a blob back. This is not the case now. Any reason for it?

For the case of a list of commands, the question is whether each line is a separate valid shell program. If not, then we need to concatenate it before processing and then split back. This is problematic. I will check if we can use something more complex than `\n` to do concatenate-and-split that our shell parser is ok with. We can also split it using a custom algorithm as we know the input and know how it can be modified, but I would prefer to avoid it.